lib/data_stores/mysql_advanced_data_store.py

```diff
@@ -0,0 +1,636 @@
+
+    if to_insert:
+      if sync:
+        transaction = self._BuildInserts(to_insert)
+        self._ExecuteInsert(transaction)
+      else:
+        with self.lock:
+          self.to_insert.extend(to_insert)
+
+  def Flush(self):
+    with self.lock:
+      to_insert = self.to_insert
+      self.to_insert = []
+
+    if to_insert:
+      transaction = self._BuildInserts(to_insert)
+      self._ExecuteInsert(transaction)
+
+  def _CountDuplicateAttributes(self, subject, predicate):
+    query = "SELECT count(*) AS total FROM aff4 WHERE \
+subject_hash=unhex(md5(%s)) AND attribute_hash=unhex(md5(%s))"
+    args = [subject, predicate]
+    result = self._ExecuteQuery(query, args)
+    return int(result[0]["total"])
+
+  def _BuildReplaces(self, values):
+    transaction = []
+    value_sets = zip(values[0::4], values[1::4], values[2::4], values[3::4])
+
+    for (subject, predicate, value, timestamp) in value_sets:
+      aff4 = {}
+      aff4["query"] = "UPDATE aff4 SET value=%s, timestamp=%s WHERE \
+subject_hash=unhex(md5(%s)) AND attribute_hash=unhex(md5(%s))"
+      aff4["args"] = [value, timestamp, subject, predicate]
+      transaction.append(aff4)
+    return transaction
+
+  def _BuildInserts(self, values):
+    subjects = {}
+    attributes = {}
+    aff4 = {}
+
+    subjects["query"] = "INSERT IGNORE INTO subjects (hash, subject) VALUES"
+    attributes["query"] = "INSERT IGNORE INTO attributes (hash, attribute) VALUES"
+    aff4["query"] = "INSERT INTO aff4 (subject_hash, attribute_hash, prefix, \
+timestamp, value) VALUES"
+
+    subjects["args"] = []
+    attributes["args"] = []
+    aff4["args"] = []
+
+    value_sets = zip(values[0::4], values[1::4], values[2::4], values[3::4])
+
+    seen = {}
+    seen["subjects"] = []
+    seen["attributes"] = []
+
+    for (subject, predicate, value, timestamp) in value_sets:
+      if subject not in seen["subjects"]:
+        subjects["args"].extend([subject, subject])
+        seen["subjects"].append(subject)
+      if predicate not in seen["attributes"]:
+        attributes["args"].extend([predicate, predicate])
+        seen["attributes"].append(predicate)
+      prefix = predicate.split(":", 1)[0]
+      aff4["args"].extend([subject, predicate, prefix, timestamp, value])
+
+    subjects["query"] += ", ".join(["(unhex(md5(%s)), %s)"] * (len(subjects["args"]) / 2))
+    attributes["query"] += ", ".join(["(unhex(md5(%s)), %s)"] * (len(attributes["args"]) / 2))
+    aff4["query"] += ", ".join(
+        ["(unhex(md5(%s)), unhex(md5(%s)), %s, %s, %s)"] * (
+            len(aff4["args"]) / 5))
+
+    return [subjects, attributes, aff4]
+
+  def _ExecuteInsert(self, transaction):
+    """Get connection from pool and execute query"""
+    for query in transaction:
+      with self.pool.GetConnection() as cursor:
+        cursor.Execute(query["query"], query["args"])
+
+  def _ResolveRegex(self, subject, predicate_regex, token=None, timestamp=None, limit=None):
+    if limit and limit == 0:
```

Comment:
I think limit = 0 normally gets treated as literally 0 results allowed while limit = None is no limit.  This line just doesn't work for returning empty results for a limit of 0 because the logic in the if is bad.  However, the calling code should prevent this from even being called with limit = 0 in the first place.